File: apps/mobile/lib/models/message.dart
enum MessageType {
  text,
  image,
  video,
  audio,
  file,
  location,
  contact,
  sticker,
  gif,
  voice,
  system,
}

enum MessageStatus {
  sent,
  delivered,
  read,
  failed,
}

class Message {
  final String id;
  final String chatId;
  final String senderId;
  final String senderName;
  final String? senderAvatar;
  final String content;
  final MessageType type;
  final MessageStatus status;
  final DateTime timestamp;
  final DateTime? updatedAt;
  final String? fileName;
  final String? fileUrl;
  final String? thumbnailUrl;
  final int? fileSize;
  final String? mimeType;
  final double? latitude;
  final double? longitude;
  final String? locationName;
  final int? duration; // 用于音频和视频
  final String? replyToId;
  final String? replyToContent;
  final bool isForwarded;
  final List<String> forwardedFrom;
  final Map<String, dynamic>? metadata;

  Message({
    required this.id,
    required this.chatId,
    required this.senderId,
    required this.senderName,
    this.senderAvatar,
    required this.content,
    required this.type,
    this.status = MessageStatus.sent,
    required this.timestamp,
    this.updatedAt,
    this.fileName,
    this.fileUrl,
    this.thumbnailUrl,
    this.fileSize,
    this.mimeType,
    this.latitude,
    this.longitude,
    this.locationName,
    this.duration,
    this.replyToId,
    this.replyToContent,
    this.isForwarded = false,
    this.forwardedFrom = const [],
    this.metadata,
  });

  Message copyWith({
    String? id,
    String? chatId,
    String? senderId,
    String? senderName,
    String? senderAvatar,
    String? content,
    MessageType? type,
    MessageStatus? status,
    DateTime? timestamp,
    DateTime? updatedAt,
    String? fileName,
    String? fileUrl,
    String? thumbnailUrl,
    int? fileSize,
    String? mimeType,
    double? latitude,
    double? longitude,
    String? locationName,
    int? duration,
    String? replyToId,
    String? replyToContent,
    bool? isForwarded,
    List<String>? forwardedFrom,
    Map<String, dynamic>? metadata,
  }) {
    return Message(
      id: id ?? this.id,
      chatId: chatId ?? this.chatId,
      senderId: senderId ?? this.senderId,
      senderName: senderName ?? this.senderName,
      senderAvatar: senderAvatar ?? this.senderAvatar,
      content: content ?? this.content,
      type: type ?? this.type,
      status: status ?? this.status,
      timestamp: timestamp ?? this.timestamp,
      updatedAt: updatedAt ?? this.updatedAt,
      fileName: fileName ?? this.fileName,
      fileUrl: fileUrl ?? this.fileUrl,
      thumbnailUrl: thumbnailUrl ?? this.thumbnailUrl,
      fileSize: fileSize ?? this.fileSize,
      mimeType: mimeType ?? this.mimeType,
      latitude: latitude ?? this.latitude,
      longitude: longitude ?? this.longitude,
      locationName: locationName ?? this.locationName,
      duration: duration ?? this.duration,
      replyToId: replyToId ?? this.replyToId,
      replyToContent: replyToContent ?? this.replyToContent,
      isForwarded: isForwarded ?? this.isForwarded,
      forwardedFrom: forwardedFrom ?? this.forwardedFrom,
      metadata: metadata ?? this.metadata,
    );
  }

  Map<String, dynamic> toMap() {
    return {
      'id': id,
      'chatId': chatId,
      'senderId': senderId,
      'senderName': senderName,
      'senderAvatar': senderAvatar,
      'content': content,
      'type': type.index,
      'status': status.index,
      'timestamp': timestamp.millisecondsSinceEpoch,
      'updatedAt': updatedAt?.millisecondsSinceEpoch,
      'fileName': fileName,
      'fileUrl': fileUrl,
      'thumbnailUrl': thumbnailUrl,
      'fileSize': fileSize,
      'mimeType': mimeType,
      'latitude': latitude,
      'longitude': longitude,
      'locationName': locationName,
      'duration': duration,
      'replyToId': replyToId,
      'replyToContent': replyToContent,
      'isForwarded': isForwarded,
      'forwardedFrom': forwardedFrom,
      'metadata': metadata,
    };
  }

  factory Message.fromMap(Map<String, dynamic> map) {
    return Message(
      id: map['id'] ?? '',
      chatId: map['chatId'] ?? '',
      senderId: map['senderId'] ?? '',
      senderName: map['senderName'] ?? '',
      senderAvatar: map['senderAvatar'],
      content: map['content'] ?? '',
      type: MessageType.values[map['type'] ?? 0],
      status: MessageStatus.values[map['status'] ?? 0],
      timestamp: DateTime.fromMillisecondsSinceEpoch(map['timestamp']),
      updatedAt: map['updatedAt'] != null
          ? DateTime.fromMillisecondsSinceEpoch(map['updatedAt'])
          : null,
      fileName: map['fileName'],
      fileUrl: map['fileUrl'],
      thumbnailUrl: map['thumbnailUrl'],
      fileSize: map['fileSize'],
      mimeType: map['mimeType'],
      latitude: map['latitude'],
      longitude: map['longitude'],
      locationName: map['locationName'],
      duration: map['duration'],
      replyToId: map['replyToId'],
      replyToContent: map['replyToContent'],
      isForwarded: map['isForwarded'] ?? false,
      forwardedFrom: List<String>.from(map['forwardedFrom'] ?? []),
      metadata: map['metadata'],
    );
  }

  @override
  String toString() {
    return 'Message(id: $id, type: $type, content: $content, timestamp: $timestamp)';
  }

  @override
  bool operator ==(Object other) {
    if (identical(this, other)) return true;
    return other is Message && other.id == id;
  }

  @override
  int get hashCode => id.hashCode;
}

File: apps/mobile/src/domain/entities/Message.ts
import type { Message as DomainMessage } from "@whatschat/shared-types";
import { toOptionalDate, toOptionalNumber } from "./valueCoercion";

export enum MessageType {
  Text = "text",
  Image = "image",
  Video = "video",
  Audio = "audio",
  File = "file",
  Location = "location",
  Contact = "contact",
  Sticker = "sticker",
  Gif = "gif",
  Voice = "voice",
  System = "system",
}

export enum MessageStatus {
  Sent = "sent",
  Delivered = "delivered",
  Read = "read",
  Failed = "failed",
}

export interface Message extends Omit<DomainMessage, "type" | "status"> {
  chatId: string;
  senderId: string;
  senderName: string;
  type: MessageType;
  status: MessageStatus;
  timestamp: Date;
  isForwarded: boolean;
  forwardedFrom: string[];
}

export class MessageEntity implements Message {
  id: string;
  chatId: string;
  senderId: string;
  senderName: string;
  senderAvatar?: string;
  content: string;
  type: MessageType;
  status: MessageStatus;
  timestamp: Date;
  updatedAt?: Date;
  fileName?: string;
  fileUrl?: string;
  thumbnailUrl?: string;
  fileSize?: number;
  mimeType?: string;
  latitude?: number;
  longitude?: number;
  locationName?: string;
  duration?: number;
  replyToId?: string;
  replyToContent?: string;
  isForwarded: boolean;
  forwardedFrom: string[];
  metadata?: Record<string, unknown>;

  constructor(data: Message) {
    this.id = data.id;
    this.chatId = data.chatId;
    this.senderId = data.senderId;
    this.senderName = data.senderName;
    this.senderAvatar = data.senderAvatar;
    this.content = data.content;
    this.type = data.type;
    this.status = data.status ?? MessageStatus.Sent;
    this.timestamp = data.timestamp;
    this.updatedAt = toOptionalDate(data.updatedAt);
    this.fileName = data.fileName;
    this.fileUrl = data.fileUrl;
    this.thumbnailUrl = data.thumbnailUrl;
    this.fileSize = toOptionalNumber(data.fileSize);
    this.mimeType = data.mimeType;
    this.latitude = data.latitude;
    this.longitude = data.longitude;
    this.locationName = data.locationName;
    this.duration = data.duration;
    this.replyToId = data.replyToId;
    this.replyToContent = data.replyToContent;
    this.isForwarded = data.isForwarded ?? false;
    this.forwardedFrom = data.forwardedFrom ?? [];
    this.metadata = data.metadata;
  }

  copyWith(updates: Partial<Message>): MessageEntity {
    return new MessageEntity({
      id: updates.id ?? this.id,
      chatId: updates.chatId ?? this.chatId,
      senderId: updates.senderId ?? this.senderId,
      senderName: updates.senderName ?? this.senderName,
      senderAvatar: updates.senderAvatar ?? this.senderAvatar,
      content: updates.content ?? this.content,
      type: updates.type ?? this.type,
      status: updates.status ?? this.status,
      timestamp: updates.timestamp ?? this.timestamp,
      updatedAt: updates.updatedAt ?? this.updatedAt,
      fileName: updates.fileName ?? this.fileName,
      fileUrl: updates.fileUrl ?? this.fileUrl,
      thumbnailUrl: updates.thumbnailUrl ?? this.thumbnailUrl,
      fileSize: updates.fileSize ?? this.fileSize,
      mimeType: updates.mimeType ?? this.mimeType,
      latitude: updates.latitude ?? this.latitude,
      longitude: updates.longitude ?? this.longitude,
      locationName: updates.locationName ?? this.locationName,
      duration: updates.duration ?? this.duration,
      replyToId: updates.replyToId ?? this.replyToId,
      replyToContent: updates.replyToContent ?? this.replyToContent,
      isForwarded: updates.isForwarded ?? this.isForwarded,
      forwardedFrom: updates.forwardedFrom ?? this.forwardedFrom,
      metadata: updates.metadata ?? this.metadata,
    });
  }

  toMap(): Record<string, unknown> {
    return {
      id: this.id,
      chatId: this.chatId,
      senderId: this.senderId,
      senderName: this.senderName,
      senderAvatar: this.senderAvatar,
      content: this.content,
      type: this.type,
      status: this.status,
      timestamp: this.timestamp.getTime(),
      updatedAt: this.updatedAt?.getTime(),
      fileName: this.fileName,
      fileUrl: this.fileUrl,
      thumbnailUrl: this.thumbnailUrl,
      fileSize: this.fileSize,
      mimeType: this.mimeType,
      latitude: this.latitude,
      longitude: this.longitude,
      locationName: this.locationName,
      duration: this.duration,
      replyToId: this.replyToId,
      replyToContent: this.replyToContent,
      isForwarded: this.isForwarded,
      forwardedFrom: this.forwardedFrom,
      metadata: this.metadata,
    };
  }

  static fromMap(map: Record<string, unknown>): MessageEntity {
    const messageTypeMap: Record<number, MessageType> = {
      0: MessageType.Text,
      1: MessageType.Image,
      2: MessageType.Video,
      3: MessageType.Audio,
      4: MessageType.File,
      5: MessageType.Location,
      6: MessageType.Contact,
      7: MessageType.Sticker,
      8: MessageType.Gif,
      9: MessageType.Voice,
      10: MessageType.System,
    };

    const messageStatusMap: Record<number, MessageStatus> = {
      0: MessageStatus.Sent,
      1: MessageStatus.Delivered,
      2: MessageStatus.Read,
      3: MessageStatus.Failed,
    };

    return new MessageEntity({
      id: (map.id as string) ?? "",
      chatId: (map.chatId as string) ?? "",
      senderId: (map.senderId as string) ?? "",
      senderName: (map.senderName as string) ?? "",
      senderAvatar: map.senderAvatar as string | undefined,
      content: (map.content as string) ?? "",
      type: messageTypeMap[(map.type as number) ?? 0] ?? MessageType.Text,
      status: messageStatusMap[(map.status as number) ?? 0] ?? MessageStatus.Sent,
      timestamp: new Date(map.timestamp as number),
      updatedAt: map.updatedAt ? new Date(map.updatedAt as number) : undefined,
      fileName: map.fileName as string | undefined,
      fileUrl: map.fileUrl as string | undefined,
      thumbnailUrl: map.thumbnailUrl as string | undefined,
      fileSize: map.fileSize as number | undefined,
      mimeType: map.mimeType as string | undefined,
      latitude: map.latitude as number | undefined,
      longitude: map.longitude as number | undefined,
      locationName: map.locationName as string | undefined,
      duration: map.duration as number | undefined,
      replyToId: map.replyToId as string | undefined,
      replyToContent: map.replyToContent as string | undefined,
      isForwarded: (map.isForwarded as boolean) ?? false,
      forwardedFrom: Array.isArray(map.forwardedFrom) ? (map.forwardedFrom as string[]) : [],
      metadata: map.metadata as Record<string, unknown> | undefined,
    });
  }

  toString(): string {
    return `Message(id: ${this.id}, type: ${this.type}, content: ${this.content}, timestamp: ${this.timestamp})`;
  }

  equals(other: Message): boolean {
    return this.id === other.id;
  }
}

